libs/libc/stdio/lib_fopencookie.c

```diff
@@ -0,0 +1,181 @@
+/****************************************************************************
+ * libs/libc/stdio/lib_fopencookie.c
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.  The
+ * ASF licenses this file to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance with the
+ * License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ ****************************************************************************/
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <nuttx/config.h>
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <unistd.h>
+#include <fcntl.h>
+#include <string.h>
+#include <assert.h>
+#include <errno.h>
+#include <nuttx/tls.h>
+
+#include "libc.h"
+
+/****************************************************************************
+ * Private Types
+ ****************************************************************************/
+
+static ssize_t freadcookie(FAR void *s, FAR char *buf, size_t nbytes);
+static ssize_t fwritecookie(FAR void *s, FAR const char *buf, size_t nbytes);
+static off_t fseekcookie(FAR void *s, off_t offset, int whence);
+static int fclosecookie(FAR void *s);
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+static FAR FILE *fopencookie_init_stream(int oflags)
+{
+  FAR FILE *stream;
+
+  stream = lib_zalloc(sizeof(FILE));
```

Comment:
stream need allocate through fs_fdopen like fopen/fdopen too.

Reply:
I cannot call fs_fdopen without valid fd.

Reply:
In the new design, FILE doesn't couple with fd, so fs_fdopen should change the first argument from fd to cookie.

Reply:
Done.